presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushdownFilterIntoRowNumber.java

```diff
@@ -0,0 +1,179 @@
+
+import static com.google.common.base.Verify.verify;
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.spi.predicate.Marker.Bound.BELOW;
+import static io.prestosql.spi.predicate.Range.range;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.sql.planner.DomainTranslator.fromPredicate;
+import static io.prestosql.sql.planner.plan.Patterns.filter;
+import static io.prestosql.sql.planner.plan.Patterns.rowNumber;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static java.lang.Math.toIntExact;
+
+public class PushdownFilterIntoRowNumber
+        implements Rule<FilterNode>
+{
+    private static final Capture<RowNumberNode> CHILD = newCapture();
+    private static final Pattern<FilterNode> PATTERN = filter().with(source().matching(rowNumber().capturedAs(CHILD)));
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+
+    public PushdownFilterIntoRowNumber(Metadata metadata)
+    {
+        this.metadata = metadata;
+        this.domainTranslator = new DomainTranslator(metadata);
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        Session session = context.getSession();
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+
+        DomainTranslator.ExtractionResult extractionResult = fromPredicate(metadata, session, node.getPredicate(), types);
+        TupleDomain<Symbol> tupleDomain = extractionResult.getTupleDomain();
+
+        RowNumberNode source = captures.get(CHILD);
+        Symbol rowNumberSymbol = source.getRowNumberSymbol();
+        OptionalInt upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
+
+        if (upperBound.isEmpty()) {
+            return Result.empty();
+        }
+
+        if (upperBound.getAsInt() <= 0) {
+            return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
+        }
+
+        int newRowCountPerPartition = upperBound.getAsInt();
+        boolean needRewriteSource = true;
+        if (source.getMaxRowCountPerPartition().isPresent()) {
+            newRowCountPerPartition = Math.min(source.getMaxRowCountPerPartition().get(), upperBound.getAsInt());
+            // No need to rewrite only source node if rowCountPerPartition results in identical.
+            needRewriteSource = source.getMaxRowCountPerPartition().get() != newRowCountPerPartition;
+        }
+
+        source = new RowNumberNode(
+                    source.getId(),
+                    source.getSource(),
+                    source.getPartitionBy(),
+                    source.isOrderSensitive(),
+                    source.getRowNumberSymbol(),
+                    Optional.of(newRowCountPerPartition),
+                    source.getHashSymbol());
```

Comment:
`RowNumberNode` is necessary anyway in the following path. For instance in L.128. Therefore, I think we can construct the source node here.

Reply:
If `needRewriteSource`, we do not need to reuse the source node. Just returning `Result.empty()` is sufficient.

Reply:
`source` is also used behind this point -- in lines 128 and 132. Those are the cases where filter predicate changes. 
They do not depend on whether the source changed or not. My suggestion is that you only create a new RowNumberNode if the `maxRowCountPerPartition` changes (which is equivalent to `needRewriteSource`). Otherwise you can use the original `source` from `CHILD` capture instead of creating a new identical instance.

Reply:
Ah, that makes sense. Thank!